### PR TITLE
Authentication and Authorization Support

### PR DESCRIPTION
**Pull request by kikkomep**

### CODE REVIEW

File: lifemonitor/auth/oauth2/server/models.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+            user_id=grant_user.get_user_id(),
+        )
+        db.session.add(item)
+        db.session.commit()
+        return code
+
+    def parse_authorization_code(self, code, client):
+        item = AuthorizationCode.query.filter_by(
+            code=code, client_id=client.client_id).first()
+        if item and not item.is_expired():
```
Comment (ilveroluca) [inline lifemonitor/auth/oauth2/server/models.py:133]: need to return something if the `if` is False

File: lifemonitor/auth/oauth2/server/models.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+
+class PasswordGrant(grants.ResourceOwnerPasswordCredentialsGrant):
+    TOKEN_ENDPOINT_AUTH_METHODS = [
+        'client_secret_basic', 'client_secret_post'
+    ]
+
+    def authenticate_user(self, username, password):
+        user = User.query.filter_by(username=username).first()
+        if user.check_password(password):
```
Comment (ilveroluca) [inline lifemonitor/auth/oauth2/server/models.py:151]: need to return something if the if is False

File: lifemonitor/auth/oauth2/server/models.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    TOKEN_ENDPOINT_AUTH_METHODS = [
+        'client_secret_basic', 'client_secret_post'
+    ]
+    INCLUDE_NEW_REFRESH_TOKEN = True
+
+    def authenticate_refresh_token(self, refresh_token):
+        item = Token.query.filter_by(refresh_token=refresh_token).first()
+        # define is_refresh_token_valid by yourself
+        # usually, you should check if refresh token is expired and revoked
+        if item and item.is_refresh_token_valid():
```
Comment (ilveroluca) [inline lifemonitor/auth/oauth2/server/models.py:171]: need to return something if the if is False

File: lifemonitor/auth/oauth2/client/models.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        for k, v, in data.items():
+            setattr(profile, k, v)
+        return profile
+
+
+class OAuthIdentity(db.Model):
+    id = db.Column(db.Integer, primary_key=True)
+    user_id = db.Column(db.Integer, db.ForeignKey(User.id), nullable=False)
+    provider_user_id = db.Column(db.String(256), nullable=False)
+    provider = db.Column(db.String(50), nullable=False)
```
[the remark this answers is not in the document]
Comment (kikkomep) [reply lifemonitor/auth/oauth2/client/models.py]: Ok, fixed via https://github.com/crs4/life_monitor/pull/21/commits/44eb92d9f63df9faed97ff943a0990225cb119b2

File: lifemonitor/auth/services.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+login_manager = LoginManager()
+login_manager.anonymous_user = Anonymous
+
+
+@login_manager.user_loader
+def load_user(user_id):
+    return User.query.get(int(user_id))
+
+
+def generate_new_api(user, scope, length=40) -> ApiKey:
```
[the remark this answers is not in the document]
Comment (kikkomep) [reply lifemonitor/auth/services.py]: Fixed by https://github.com/crs4/life_monitor/pull/21/commits/1c44a49f518c090246e3415cb2678615dd56ca18

File: lifemonitor/auth/oauth2/server/models.py (hunk 1 above)
Comment (kikkomep) [reply lifemonitor/auth/oauth2/server/models.py]: Fixed by https://github.com/crs4/life_monitor/pull/21/commits/3745ae6897a194437fad2240bcde0df926e85353

File: lifemonitor/auth/oauth2/server/models.py (hunk 2 above)
Comment (kikkomep) [reply lifemonitor/auth/oauth2/server/models.py]: Fixed by https://github.com/crs4/life_monitor/pull/21/commits/7c480e1a2b9cb66bef55bb36025153b4c31cd7e3

File: lifemonitor/auth/oauth2/server/models.py (hunk 3 above)
Comment (kikkomep) [reply lifemonitor/auth/oauth2/server/models.py]: Fixed by https://github.com/crs4/life_monitor/pull/21/commits/a7381aa1f40c13f9811e173829f2da3245fef762

**Pull request merged**